e2e/nomostest/reset.go

```diff
@@ -0,0 +1,513 @@
+
+	return nil
+}
+
+func protectedNamespaceList() []string {
+	list := make([]string, 0, len(differ.SpecialNamespaces))
+	for ns := range differ.SpecialNamespaces {
+		list = append(list, ns)
+	}
+	return list
+}
+
+type resetRecord struct {
+	Objects          map[types.NamespacedName]struct{}
+	ManagedObjects   map[types.NamespacedName]struct{}
+	ObjectNamespaces map[string]struct{}
+}
+
+func newResetRecord() *resetRecord {
+	return &resetRecord{
+		Objects:          make(map[types.NamespacedName]struct{}),
+		ManagedObjects:   make(map[types.NamespacedName]struct{}),
+		ObjectNamespaces: make(map[string]struct{}),
+	}
+}
+
+// ResetRootSyncs cleans up one or more RootSyncs and all their managed objects.
+// Use this for cleaning up RootSyncs in tests that use delegated control.
+func ResetRootSyncs(nt *NT, rsList []v1beta1.RootSync) error {
+	if len(rsList) == 0 {
+		return nil
+	}
+
+	nt.T.Logf("[RESET] Deleting RootSyncs (%d)", len(rsList))
+
+	record := newResetRecord()
+
+	for _, item := range rsList {
+		rs := &item
+		rsNN := client.ObjectKeyFromObject(rs)
+		record.Objects[rsNN] = struct{}{}
+		record.ObjectNamespaces[rsNN.Namespace] = struct{}{}
+
+		if manager, found := rs.GetAnnotations()[string(metadata.ResourceManagerKey)]; found {
+			record.ManagedObjects[rsNN] = struct{}{}
+			nt.T.Logf("[RESET] RootSync %s managed by %q", rsNN, manager)
+			if !IsDeletionPropagationEnabled(rs) {
+				// If you go this error, make sure your test cleanup ensures
+				// that the managed RootSync has deletion propagation enabled.
+				return errors.Errorf("RootSync %s managed by %q does NOT have deletion propagation enabled: test reset incomplete", rsNN, manager)
+			}
+			continue
+		}
+
+		// Enable deletion propagation, if not enabled
+		if EnableDeletionPropagation(rs) {
+			nt.T.Logf("[RESET] Enabling deletion propagation on RootSync %s", rsNN)
+			if err := nt.Update(rs); err != nil {
+				return err
+			}
+			if err := WatchObject(nt, kinds.RootSyncV1Beta1(), rs.Name, rs.Namespace, []Predicate{
+				HasFinalizer(metadata.ReconcilerFinalizer),
+			}); err != nil {
+				return err
+			}
+		}
+
+		// Print reconciler logs in case of failure.
+		// This ensures the logs are printed, even if the reconciler is deleted.
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		go TailReconcilerLogs(ctx, nt, RootReconcilerObjectKey(rsNN.Name))
+
+		// DeletePropagationBackground is required when deleting RSyncs with
+		// dependencies that have owners references. Otherwise the reconciler
+		// and dependenencies will be garbage collected before the finalizer
+		// can delete the managed resources.
+		// TODO: Remove explicit Background policy after the reconciler-manager finalizer is added.
+		nt.T.Logf("[RESET] Deleting RootSync %s", rsNN)
+		if err := nt.Delete(rs, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+			return err
+		}
+	}
+	tg := taskgroup.New()
+	for _, item := range rsList {
```

Comment:
I think generally this should result in all of the `RootSync` objects being deleted, but I'm trying to think of unexpected corner cases. For example:
- a `RootSync` in the input is managed by a `RootSync` that is not included in the input
- a `RootSync` in the input is managed by a `RepoSync`

It might be the case that neither of these scenarios ever happen, but just wanted to bring up this consideration

Reply:
> a RootSync in the input is managed by a RootSync that is not included in the input

Mmm... Yeah... decoupling the List did make it risky...

The List in Reset pulls all the RootSyncs that have the test label. So it should be all of them. But it's possible that someone could call ResetRootSyncs with some other set of RootSyncs and it might break.

My goal with the reset code wasn't to cover 100% of possible edge cases tho, only the ones the tests _should_ be using. If a test violates the expectations and Reset breaks, the Clean should fix it because it's more aggressive and thorough (and slower).

> a RootSync in the input is managed by a RepoSync

I think the webhook rejects requests to create RepoSyncs in the config-management-system namespace.  You can have RootSyncs manage RootSyncs, but not RepoSyncs managing RootSyncs.

Reply:
If you can think of a quick way to detect or mitigate those tho, I can add them in...

Reply:
I think it would be sufficient to just include a cautionary warning in the docstring to inform future adventurers. e.g. something like
```
this function assumes all RootSyncs in the cluster are passed in as input
```

Reply:
Assuming CI is happy later, I'll make some follow up PRs to do minor improvements to the reset code. But I don't want to do them now in case we need to revert again.